### PR TITLE
Upgrade in-memory and on-disk state representation with tree states

### LINKED ISSUE #2371: Implement consensus context optimisation
## Description

Throughout state processing, we quite often make use of a pattern where we pass in a pre-computed value in order to avoid recomputing it, e.g. the `state_root` passed to `per_slot_processing` or the `total_active_balance` passed to various rewards processing functions. There are more opportunities for such optimisations, including the proposer index (https://github.com/sigp/lighthouse/issues/598) and other properties of the validator set (`unslashed_participating_indices` for the 3 different flags).

My idea for neatly generalising this pattern and enabling more of these optimisations is to add a single context struct to contain these values, and pass it around between all relevant state transition functions:

```rust
pub struct ConsensusContext<T: EthSpec> {
    /// Slot to act as an identifier/safeguard
    slot: Slot,
    state_root: Option<Hash256>,
    proposer_index: Option<u64>,
    ...
}
```

The signature of `per_slot_processing` would change as follows:

```rust
pub fn per_slot_processing<T: EthSpec>(
    state: &mut BeaconState<T>,
    context: &mut ConsensusContext<T>,
    spec: &ChainSpec,
) -> Result<Option<EpochProcessingSummary>, Error>;
```

The context would have constructors/builders for making a context with 0 or more pre-computed values. If you have the `state_root` handy, you feed it in, or if you have the proposer index from the `BeaconChain`'s proposer index cache, you supply that.

```rust
let mut ctxt = ConsensusContext::new(slot)
    .set_state_root(state_root)
    .set_proposer_index(proposer_index);
```

Then in the consensus code when these values are required they are fetched via memoising accessors that either a) use the cached value if it exists, or b) compute the value from scratch and add it to the context for future use.

```rust
impl<T: EthSpec> ConsensusContext<T> {
    pub fn get_beacon_proposer_index(
        &mut self,
        state: &BeaconState<T>,
        spec: &ChainSpec
    ) -> Result<usize, Error> {
        if let Some(proposer_index) = self.proposer_index {
            return Ok(proposer_index);
        }
        let proposer_index = state.get_beacon_proposer_index(self.slot, spec)?;
        self.proposer_index = Some(proposer_index)
        Ok(proposer_index)
    }
}
```

## Alternatives

* As an alternative we could instead add these optional memoising caches to the `BeaconState`, but I think that would be messier, and would introduce cache invalidation bugs that don't exist with an ephemeral cache like the `ConsensusContext`
* Plumb parameters around manually. Also messier, introduces friction for new optimisations, no memoisation logic.

### LINKED ISSUE #2805: Improve block processing performance during re-org
## Description

Consider the following re-org that frustrates Lighthouse's attempts to process blocks quickly:

Let `n` be a slot on an epoch boundary (`n % 32 == 0`).

1. Immediately prior to slot `n` the preemptive state advance occurs as normal
2. The block from slot `n` arrives super late (12s+), consuming the advanced state
3. The block from slot `n + 1` arrives on time, but builds upon the parent at slot `n -1`. It's going to be super slow to process because its parent state is missing from the cache, meaning:
    a) We need to load the full state for slot `n - 1` from disk (a few hundred ms)
    b) We need to transition that state through an epoch boundary (200ms)
    c) We need to store the state for slot `n` on disk. It is different from the slot `n` slot with block `n` applied, and presently we store every epoch boundary state

## Example

Here's an instance of this behaviour that I observed at slot `n=2485472` on mainnet, resulting in block processing taking 2.5s instead of the usual 80ms (median) or 456ms (99th percentile) (metrics from https://github.com/sigp/lighthouse-metrics/pull/31).

```
Nov 11 16:55:03.815 WARN Beacon chain re-org                     reorg_distance: 1, new_slot: 2485473, new_head: 0xb17f…a572, new_head_parent: 0x0f98…9b22, previous_slot: 2485472, previous_head: 0x1c4d…c94b, service: beacon
Nov 11 16:55:03.818 DEBG Delayed head block                      set_as_head_delay: Some(222.219889ms), imported_delay: Some(2.545278411s), observed_delay: Some(2.051036927s), block_delay: 4.818535227s, slot: 2485473, proposer_index: 52065, block_root: 0xb17fe52ce55315713a9e3eb28858a1a53039daf9e1f6406aa2c8d0d8ae11a572, service: beacon
```

Even though the block arrived on time, taking 2.5s to process it meant that any attestations at this slot would have missed (if running on this node).

## Additional Info

It should be noted that this behaviour should be quite rare, due to the infrequency of re-orgs and late blocks on mainnet (at worst ~4% of blocks are late, with very few being 12s+ late). However if proposer boosting is adopted we may see more re-orgs of this type, where a proposer intentionally orphans the previous block despite it having been published.

### LINKED ISSUE #2806: Persistent copy-on-write beacon states
## Description

One strategy for improving block processing times during re-org (https://github.com/sigp/lighthouse/issues/2805) that I've been thinking about for a while would be to make it _extremely cheap_ to clone and duplicate `BeaconState`s. Currently a lot of our caches and optimisations dance around the fact that `BeaconState`s are huge 30-50MB objects that take 200ms+ to clone. Building a tree hash cache from scratch also takes ~1s.

Using persistent data structures that are copy-on-write, we could exploit the large amount of data shared between states at nearby slots. If each `BeaconState` were represented as a tree with `Arc` nodes, then we could keep a large number of them in memory, and use these to accelerate block processing. In smooth network conditions we could probably keep every state since finalization in memory: with and without blocks applied.

On the implementation side, we could revive my old implementation of the "validator tree" from EthDenver 2020, which lives in this defunct branch: https://github.com/michaelsproul/lighthouse/tree/validator-tree-ethdenver. It was inspired by talks with Proto about [Remerkleable](https://github.com/protolambda/remerkleable).

This issue is pretty vague currently, but I'll add to it as my thinking on it develops (it also links in to using some sort of diff-based representation to store states on disk).

### LINKED ISSUE #2944: Snapshots cloned during sync
## Description

Since https://github.com/sigp/lighthouse/pull/2832 Lighthouse will clone a snapshot from the cache whenever there's a skipped slot. This is designed to help during live block processing, but _also_ gets applied during sync where it isn't very useful.

## Present Behaviour

The condition here trips regardless of whether or not Lighthouse is syncing:

https://github.com/sigp/lighthouse/blob/79db2d4deb6a47947699d8a4a39347c19ee6e5d6/beacon_node/beacon_chain/src/snapshot_cache.rs#L256-L258

## Expected Behaviour

Lighthouse should always move the snapshot from the cache during sync in order to prevent unnecessary cloning and extra memory usage.

### LINKED ISSUE #3208: Compress and linearise beacon blocks on disk
## Description

Beacon blocks alone now consume around 80GB of disk space on mainnet.

I think we could be more efficient with storage by using compression. Recently I'm particularly fond of `zstd` compression, which is fast and could probably offer a 2-4x reduction.

The complicating factor is that compressing blocks requires a database schema upgrade, as well as a moderately sophisticated migration to copy and compress all the existing blocks.

### LINKED ISSUE #3505: Store pubkey cache decompressed on disk
## Description

Currently Lighthouse takes a long time to start up while it reloads the public key cache from disk.

The fundamental reason for this is that it stores the BLS pubkeys in their compressed form and then has to decompress every one of them for the in-memory cache:

https://github.com/sigp/lighthouse/blob/18c61a5e8be3e54226a86a69b96f8f4f7fd790e4/beacon_node/beacon_chain/src/validator_pubkey_cache.rs#L166-L169

Just changing that `PublicKeyBytes` to `PublicKey` would not be sufficient, because the two types share the same compressed SSZ representation:

https://github.com/sigp/lighthouse/blob/18c61a5e8be3e54226a86a69b96f8f4f7fd790e4/crypto/bls/src/macros.rs#L29-L50

https://github.com/sigp/lighthouse/blob/18c61a5e8be3e54226a86a69b96f8f4f7fd790e4/crypto/bls/src/impls/blst.rs#L121-L124

So we would need to create new methods on `bls::GenericPublicKey` like `serialize_uncompressed` and `deserialize_uncompressed` which use the 96-byte serialization of the `PublicKey` rather than the 48-byte compressed serialization. For `blst`, we can use the [`PublicKey::serialize`](https://docs.rs/blst/0.3.10/blst/min_pk/struct.PublicKey.html#method.serialize) function. For `milagro` we can use [`PublicKey::as_uncompressed_bytes`](https://github.com/sigp/milagro_bls/blob/421aa3af8a0618a46144560389debbe7e56609db/src/keys.rs#L159-L162). Both libraries also include deserialisation functions acting on the uncompressed bytes.

Once these methods are in place in the `bls` wrapper lib, we can call them from the `DatabasePubkey` wrapper. We'll need two versions of `DatabasePubkey` in order to implement the schema migration from compressed keys to uncompressed keys.

By storing the bytes uncompressed we'll double the disk space required for the pubkey cache from 500k * 48 bytes = 24 MB to 48 MB. Given the overall size of the database I believe this is an acceptable trade-off. We could attempt to apply a faster compression algorithm like `zstd` to the data, but seeing as the input is pseudo-random bytes I suspect this naive compression would be ineffective.

## Version

Lighthouse v3.0.0

### PR DESCRIPTION
**Pull request by michaelsproul**
This is an umbrella PR for fundamental changes to Lighthouse's state representation.

The plan is to merge smaller PRs into this one, and then merge it to `unstable` without squashing.

## TODO List

There are few things that need fixing on this branch before it's ready for primetime:

- [x] Remove recursion from `get_state`. The recursion is liable to overflow the stack, so we should switch to an iterative algorithm.
- [x] Fix the validator monitor
- [x] Clean up the state processing optimisations
- [x] Fuzz the state processing optimisations
- [x] Add or fix state processing optimisations for phase0 so that block replay isn't too slow
- [x] Add new metrics for `jemalloc`
- [x] Resolve all remaining `FIXME(sproul)`s in the code
- [x] Re-do the optimisation that was reverted in #3658

## Issues Addressed

Closes #2806
Closes #2805
Closes #2944
Closes #2371
Closes #3208
Closes #3505

### CODE REVIEW

File: database_manager/src/lib.rs [hunk 1]
```diff
@@ -437,6 +439,57 @@ pub fn migrate_db<E: EthSpec>(
...
+    let second = clap_utils::parse_required(cli_args, "second")?;
+
+    Ok(DiffConfig { first, second })
+}
+
+pub fn diff<E: EthSpec>(diff_config: &DiffConfig, log: Logger) -> Result<(), Error> {
+    use ssz::{Decode, Encode};
+    use std::fs::File;
+    use std::io::Read;
+    use store::StoreConfig;
```
Comment (dapplion) [inline database_manager/src/lib.rs]: Should be global imports?

File: database_manager/src/lib.rs [hunk 2]
```diff
@@ -437,6 +439,57 @@ pub fn migrate_db<E: EthSpec>(
...
+
+    for (i, new_balance) in second.iter().enumerate() {
+        let old_balance = first.get(i).copied().unwrap_or(0);
+        let diff = new_balance.wrapping_sub(old_balance);
+        diff_balances.push(diff);
+    }
+
+    let diff_ssz_bytes = diff_balances.as_ssz_bytes();
+    let config = StoreConfig::default();
+    let compressed_diff_bytes = config.compress_bytes(&diff_ssz_bytes).unwrap();
```
Comment (dapplion) [inline database_manager/src/lib.rs]: Unwraps

File: beacon_node/store/src/iter.rs [hunk 3]
```diff
@@ -377,131 +378,3 @@ fn slot_of_prev_restore_point<E: EthSpec>(current_slot: Slot) -> Slot {
     let slots_per_historical_root = E::SlotsPerHistoricalRoot::to_u64();
     (current_slot - 1) / slots_per_historical_root * slots_per_historical_root
 }
-
-#[cfg(test)]
-mod test {
```
Comment (dapplion) [inline beacon_node/store/src/iter.rs:382]: Intentional delete? `BlockRootsIterator` still exists

File: beacon_node/store/src/leveldb_store.rs [hunk 4]
```diff
@@ -26,6 +26,7 @@ impl<E: EthSpec> LevelDB<E> {
         let mut options = Options::new();
 
         options.create_if_missing = true;
+        options.write_buffer_size = Some(512 * 1024 * 1024);
```
Comment (dapplion) [inline beacon_node/store/src/leveldb_store.rs:28]: Add comment justifying the increased value

File: consensus/state_processing/src/common/initiate_validator_exit.rs [hunk 5]
```diff
@@ -8,10 +8,10 @@ pub fn initiate_validator_exit<T: EthSpec>(
...
     spec: &ChainSpec,
 ) -> Result<(), Error> {
-    // Return if the validator already initiated exit
-    if state.get_validator(index)?.exit_epoch != spec.far_future_epoch {
-        return Ok(());
-    }
+    // We do things in a slightly different order to the spec here. Instead of immediately checking
+    // whether the validator has already exited, we instead prepare the exit cache and compute the
+    // cheap-to-calculate values from that. *Then* we look up the validator a single time in the
+    // validator tree (expensive), make the check and mutate as appropriate.
```
Comment (dapplion) [inline consensus/state_processing/src/common/initiate_validator_exit.rs]: Yes, but why?

File: consensus/state_processing/src/per_block_processing/process_operations.rs [hunk 6]
```diff
@@ -145,32 +146,36 @@ pub mod altair_deneb {
...
-                    proposer_reward_numerator.safe_add_assign(
-                        get_base_reward(state, index, base_reward_per_increment, spec)?
-                            .safe_mul(weight)?,
-                    )?;
-
-                    if flag_index == TIMELY_TARGET_FLAG_INDEX {
+                let epoch_participation = state.get_epoch_participation_mut(
+                    data.target.epoch,
+                    ctxt.previous_epoch,
+                    ctxt.current_epoch,
```
Comment (dapplion) [inline consensus/state_processing/src/per_block_processing/process_operations.rs]: Does this optimization make a difference? Introduced in https://github.com/sigp/lighthouse/commit/2df714e2cddf1131f6381e42fdf355eae1a85be1

File: consensus/types/src/beacon_state.rs [hunk 7]
```diff
@@ -1702,103 +1917,63 @@ impl<T: EthSpec> BeaconState<T> {
...
+                .map_or(false, VList::has_pending_updates)
+            || self
+                .previous_epoch_participation()
+                .map_or(false, VList::has_pending_updates)
+            || self
+                .current_epoch_participation()
+                .map_or(false, VList::has_pending_updates)
+            || self
+                .inactivity_scores()
+                .map_or(false, VList::has_pending_updates)
```
Comment (dapplion) [inline consensus/types/src/beacon_state.rs]: Forgetting to add a field in a new fork here may cause either a panic or data corruption. This function could be generated by a macro?

File: consensus/types/src/beacon_state.rs [hunk 8]
```diff
@@ -1836,10 +2011,165 @@ impl<T: EthSpec> BeaconState<T> {
...
+                    |_, self_field, base_field| { self_field.rebase_on(base_field) }
+                );
+            }
+            (Self::Capella(self_inner), Self::Capella(base_inner)) => {
+                bimap_beacon_state_capella_tree_list_fields!(
+                    self_inner,
+                    base_inner,
+                    |_, self_field, base_field| { self_field.rebase_on(base_field) }
+                );
+            }
```
Comment (dapplion) [inline consensus/types/src/beacon_state.rs:2062]: Not using Deneb's `bimap_beacon_state_deneb_tree_list_fields`

File: database_manager/src/lib.rs (hunk 1 above)
Comment (michaelsproul) [reply database_manager/src/lib.rs]: Deleted in https://github.com/sigp/lighthouse/pull/5171

File: beacon_node/store/src/hot_cold_store.rs [hunk 9]
```diff
@@ -314,7 +346,20 @@ impl<E: EthSpec> HotColdDB<E, LevelDB<E>, LevelDB<E>> {
...
+                .check_compatibility(&disk_config, &split, anchor.as_ref())?;
+
+            // Inform user if hierarchy config is changing.
+            if db.config.hierarchy_config != disk_config.hierarchy_config {
+                info!(
+                    db.log,
+                    "Updating historic state config";
+                    "previous_config" => ?disk_config.hierarchy_config,
+                    "new_config" => ?db.config.hierarchy_config,
+                );
```
Comment (dapplion) [inline beacon_node/store/src/hot_cold_store.rs:360]: @michaelsproul I recall you commenting changing the `hierarchy_config` is not possible without a re-sync. Should this comment state so?

Comment (michaelsproul) [reply beacon_node/store/src/hot_cold_store.rs:360]: Well if you see this comment then it has already worked, so not sure if it's helpful at this stage.

You can change it by running `prune-states` and then restarting with the new config. I think it probably makes sense to just have that in the docs.

File: consensus/state_processing/src/common/initiate_validator_exit.rs (hunk 5 above)
Comment (michaelsproul) [reply consensus/state_processing/src/common/initiate_validator_exit.rs]: Addressed in https://github.com/sigp/lighthouse/pull/5386

**Comment by michaelsproul** [comment]
Closing now that v6.0.0 is out with tree-states proper. There are still a few changes from OG `tree-states` that we are planning to merge, but this PR has served its purpose.

**Pull request closed without merging**